Lib/test/test_pyrepl/test_interact.py

```diff
@@ -273,3 +274,28 @@ def test_incomplete_statement(self):
         code = "if foo:"
         console = InteractiveColoredConsole(namespace, filename="<stdin>")
         self.assertTrue(_more_lines(console, code))
+
+
+class TestWarnings(unittest.TestCase):
+    def test_pep_765_warning(self):
+        """
+        Test that a SyntaxWarning emitted from the
+        AST optimizer is only shown once in the REPL.
+        """
+        # gh-131927
+        console = InteractiveColoredConsole()
+        code = dedent("""\
+        def f():
+            try:
+                return 1
+            finally:
+                return 2
+        """)
+
+        with warnings.catch_warnings(record=True) as caught:
+            warnings.simplefilter("default")
+            console.runsource(code)
+
+        count = sum("'return' in a 'finally' block" in str(w.message)
```

Comment:
You can use `.count()` here.

"'return' in a 'finally' block" may be an error in future versions. Maybe use some more long living warnings, like `assert(1, 'message')` or `1 is 1`? BTW, they are currently emitted 3 times, not 2.

Reply:
> You can use .count() here.

Hmm I don't think I can, unless I want to match the warning message exactly?

> "'return' in a 'finally' block" may be an error in future versions. Maybe use some more long living warnings, like assert(1, 'message') or 1 is 1?

I can't if I want to test the new repl. The PEP 765 warning is the only one emitted from the ast optimizer which is needed to trigger this behaviour in the repl. However this does change the behaviour of all warnings that use `_PyErr_EmitSyntaxWarning` so I also added tests to `test_compile` for it.